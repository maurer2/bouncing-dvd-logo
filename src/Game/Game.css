.game {
  width: 100vw;
  height: 100vh;
  filter: opacity(1);
  transition: filter 0.25s;
}

.game--is-paused {
  filter: opacity(0.25);
}
