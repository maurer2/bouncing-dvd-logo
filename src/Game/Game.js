import React, { Component } from 'react';

import uid from 'uid';
import debounce from 'lodash.debounce';
import styled from 'styled-components/macro';
import { Loop } from 'react-game-kit';

import Playfield from '../Playfield/Playfield';

const GameWrapper = styled.div`
  width: 100vw;
  height: 100vh;
  transition: filter 0.15s ease-in-out;
  cursor: pointer;

  ${props => (props.isPaused ? 'filter: opacity(0.25)' : 'filter: opacity(1)')};
`;

class Game extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isRunning: true,
      key: uid(4), // needed for reinitializing
      isFirstLoad: true,
    };

    this.reset = this.reset.bind(this);
    this.togglePlayState = this.togglePlayState.bind(this);
    this.handleResize = debounce(this.handleResize.bind(this), 500);
  }

  componentDidMount() {
    if (window.ResizeObserver === undefined) {
      return;
    }

    // Register observer
    const gameResizeObserver = new window.ResizeObserver((entries) => {
      const gameHasResized = entries.some(entry => entry.target === this.wapper);

      if (this.state.isFirstLoad) {
        this.setState({ isFirstLoad: false });

        return;
      }

      if (gameHasResized) {
        this.handleResize();
      }
    });

    gameResizeObserver.observe(this.wapper);
  }

  togglePlayState() {
    this.setState(previousState => ({ isRunning: !previousState.isRunning }));
  }

  reset() {
    this.setState({ key: uid(4) });
  }

  handleResize() {
    this.reset();
  }

  render() {
    return (
      <Loop>
        <GameWrapper
          isPaused={ !this.state.isRunning }
          onClick={ this.togglePlayState }
          ref={ (element) => { this.wapper = element; }}
        >
          <Playfield isPaused={ !this.state.isRunning } key={ this.state.key } />
        </GameWrapper>
      </Loop>
    );
  }
}

export default Game;
