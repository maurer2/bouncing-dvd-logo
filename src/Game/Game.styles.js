import styled from 'styled-components/macro';

// eslint-disable-next-line import/prefer-default-export
export const GameWrapper = styled.div`
  width: 100vw;
  height: 100vh;
  transition: filter 0.15s ease-in-out;
  cursor: pointer;
  color: white;

  ${(props) => (props.isPaused ? 'filter: opacity(0.25)' : 'filter: opacity(1)')};
`;
