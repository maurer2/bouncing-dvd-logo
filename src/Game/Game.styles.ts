import styled from 'styled-components';

export const GameWrapper = styled.div`
  position: relative;
  width: 100vw;
  height: 100vh;
  color: white;
  cursor: pointer;
  transition: filter 0.15s ease-in-out;
  overflow: hidden;
  outline: none;
`;

export const PauseButton = styled.button`
  position: absolute;
  inset: 0;
  appearance: none;
  background: 0;
  padding: 0;
  border: 0;
  outline: none;
  text-indent: 9999px;
  overflow: hidden;
`;
