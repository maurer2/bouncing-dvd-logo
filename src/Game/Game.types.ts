import type { PropsWithChildren } from 'react';

export type GameProps = {
  children?: PropsWithChildren['children'];
};
