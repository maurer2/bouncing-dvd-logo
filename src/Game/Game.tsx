import type { FC, KeyboardEvent, ReactElement, PropsWithChildren } from 'react';
import React, { useRef, useEffect, useCallback } from 'react';
import { StyleSheetManager } from 'styled-components';
import { useDispatch, useSelector } from 'react-redux';

import SoundToggle from '../SoundToggle/SoundToggle';
import PlayingField from '../Playingfield/Playingfield';
import { toggleSound, togglePlayState } from '../Store/actionCreators';
import { getSoundState, getPlayState, getIsPlayingSoundState } from '../Store/selectors';
import type { Dispatch } from '../Store/types';
import SoundPlayer from '../SoundPlayer/SoundPlayer';

import * as Styles from './Game.styles';
import type * as Types from './Game.types';

const Game: FC<Readonly<PropsWithChildren<Types.GameProps>>> = (): ReactElement => {
  const dispatch: Dispatch = useDispatch();
  const soundIsDisabled: boolean = useSelector(getSoundState);
  const isPaused: boolean = useSelector(getPlayState);
  const isPlayingSound: boolean = useSelector(getIsPlayingSoundState);
  const pauseButtonDomElement = useRef<HTMLButtonElement | null>(null);

  const handleClick = (): void => {
    dispatch(togglePlayState());
  };
  const handleInput = (event: KeyboardEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    const observedKeys = [' ', 'k']; // " " === spacebar

    if (observedKeys.includes(event.key.toLowerCase())) {
      dispatch(togglePlayState());
    }
  };

  const toggleSoundCB = useCallback(() => {
    dispatch(toggleSound());
  }, [dispatch]);

  // autofocus for pause button
  useEffect(() => {
    pauseButtonDomElement?.current?.focus();
  }, []);

  return (
    <StyleSheetManager disableVendorPrefixes>
      <Styles.GameWrapper data-testid="game">
        <PlayingField />
        <Styles.PauseButton
          tabIndex={0}
          onClick={handleClick}
          onKeyUp={handleInput}
          ref={pauseButtonDomElement}
          aria-label={isPaused ? 'Unpause' : 'Pause'}
          data-testid="pausebutton"
        />
        <SoundToggle
          soundIsDisabled={soundIsDisabled}
          toggleSound={toggleSoundCB}
        />
        <SoundPlayer shouldTriggerSound={isPlayingSound} />
      </Styles.GameWrapper>
    </StyleSheetManager>
  );
};

export default Game;
