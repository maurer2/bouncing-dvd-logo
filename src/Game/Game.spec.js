import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import Game from './Game';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('Game', () => {
  const wrapper = shallow(<Game />);

  test('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
