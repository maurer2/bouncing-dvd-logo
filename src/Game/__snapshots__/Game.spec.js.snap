// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should match snapshot 1`] = `
<Loop>
  <GameWrapper
    isPaused={false}
    onClick={[Function]}
  >
    <Playfield
      isPaused={false}
      key="test"
    />
  </GameWrapper>
</Loop>
`;
